src/transformers/generation/tf_utils.py

```diff
@@ -3066,16 +3066,20 @@ def beam_search_cond_fn(
             not_max_length_yet = cur_len < max_length
 
             # 2. can the new beams still improve?
-            best_running_score = running_scores[:, :1] / (max_length**length_penalty)
+            best_running_score = running_scores[:, :1] / tf.cast(cur_len, dtype=running_scores.dtype) ** length_penalty
             worst_finished_score = tf.where(
                 is_sent_finished, tf.math.reduce_min(scores, axis=1, keepdims=True), -1.0e9
             )
             improvement_still_possible = tf.math.reduce_all(worst_finished_score < best_running_score)
 
             # 3. is there still a beam that has not finished?
-            still_open_beam = ~(tf.math.reduce_all(is_sent_finished) & early_stopping)
+            # still_open_beam = ~(tf.math.reduce_all(is_sent_finished) & early_stopping)
+            still_open_beam = ~(tf.math.reduce_all(is_sent_finished))
 
-            return not_max_length_yet & (still_open_beam | improvement_still_possible)
+            _early_stopping = tf.constant(early_stopping > 0, dtype=tf.bool)
+
+            # return not_max_length_yet & (still_open_beam | improvement_still_possible)
+            return not_max_length_yet & (still_open_beam | (~_early_stopping & improvement_still_possible))
```

Comment:
The method `beam_search_cond_fn` corresponds to [BeamHypotheses.is_done](https://github.com/huggingface/transformers/blob/3be028bc9d4b2cce9539b940f17052f333284684/src/transformers/generation/beam_search.py#L884) in our PyTorch generation code (despite the meaning is reversed: generation done v.s. not done).

- In **current** `main` branch, the logic here is
  ```python
  still_open_beam = ~(tf.math.reduce_all(is_sent_finished) & early_stopping)
  return not_max_length_yet & (still_open_beam | improvement_still_possible)
  ```
  When `early_stopping` is `False`, `still_open_beam` will be `True` and the return value becomes `True` (if
`not_max_length_yet` is `True`) - i.e. `it should continue the generation`

- However, in [BeamHypotheses.is_done](https://github.com/huggingface/transformers/blob/3be028bc9d4b2cce9539b940f17052f333284684/src/transformers/generation/beam_search.py#L884), if `early_stopping` is `False` (and if ), it will compare the scores
https://github.com/huggingface/transformers/blob/3be028bc9d4b2cce9539b940f17052f333284684/src/transformers/generation/beam_search.py#L895-L897
and in the code snippet for #18149, it returns `True` for `is_done` after 5 or 6 generation steps - i.e. `it should **NOT** continue the generation`

The above suggests:
  - The main issue in `TFMarian` super slow generation comes from the condition around `early_stopping`
  - With the changes in this PR, it could generate quickly just as the `Marian`

Reply:
However, one thing I don't understand very well is:

this part `len(self) < self.num_beams` in [BeamHypotheses.is_done](https://github.com/huggingface/transformers/blob/3be028bc9d4b2cce9539b940f17052f333284684/src/transformers/generation/beam_search.py#L890)
https://github.com/huggingface/transformers/blob/3be028bc9d4b2cce9539b940f17052f333284684/src/transformers/generation/beam_search.py#L890

v.s.

`tf.math.reduce_all(is_sent_finished)` and/or `not_max_length_yet` in `beam_search_cond_fn`.

It doesn't seem 100% equivalent conditions.